src/analysis/processing/qgsalgorithmrandomextract.cpp

```diff
@@ -104,37 +109,106 @@ QVariantMap QgsRandomExtractAlgorithm::processAlgorithm( const QVariantMap &para
-  std::generate( fids.begin(), fids.end(), bind( fidsDistribution, mersenneTwister ) );
+  std::mt19937 mersenneTwister( randomDevice() );
 
-  QHash< QgsFeatureId, int > idsCount;
-  for ( const QgsFeatureId id : fids )
+  if ( !noDuplicates )
   {
-    if ( feedback->isCanceled() )
+    const std::uniform_int_distribution<int> fidsDistribution( 1, count );
+
+    QVector< QgsFeatureId > fids( number );
+    std::generate( fids.begin(), fids.end(), bind( fidsDistribution, mersenneTwister ) );
+
+    QHash< QgsFeatureId, int > idsCount;
+    for ( const QgsFeatureId id : fids )
     {
-      break;
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+
+      idsCount[ id ] += 1;
     }
 
-    idsCount[ id ] += 1;
-  }
+    const QgsFeatureIds ids = qgis::listToSet( idsCount.keys() );
+    QgsFeatureIterator fit = source->getFeatures( QgsFeatureRequest().setFilterFids( ids ), QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks );
 
-  const QgsFeatureIds ids = qgis::listToSet( idsCount.keys() );
-  QgsFeatureIterator fit = source->getFeatures( QgsFeatureRequest().setFilterFids( ids ), QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks );
+    QgsFeature f;
+    while ( fit.nextFeature( f ) )
+    {
+      if ( feedback->isCanceled() )
+      {
+        break;
+      }
+
+      const int count = idsCount.value( f.id() );
+      for ( int i = 0; i < count; ++i )
+      {
+        if ( !sink->addFeature( f, QgsFeatureSink::FastInsert ) )
+          throw QgsProcessingException( writeFeatureError( sink.get(), parameters, QStringLiteral( "OUTPUT" ) ) );
+      }
+    }
+  }
 
-  QgsFeature f;
-  while ( fit.nextFeature( f ) )
+  // No duplicates
+  else
   {
-    if ( feedback->isCanceled() )
+    // Build a list of all feature ids
+    QgsFeatureIterator fit = source->getFeatures( QgsFeatureRequest()
+                             .setFlags( QgsFeatureRequest::NoGeometry )
+                             .setNoAttributes() );
+    std::vector< QgsFeatureId > allFeats;
+    allFeats.reserve( count );
+    QgsFeature f;
+    while ( fit.nextFeature( f ) )
+    {
+      if ( feedback->isCanceled() )
+        break;
+      allFeats.push_back( f.id() );
+    }
+
+    std::uniform_int_distribution<int> fidsDistribution;
+
+    int nb = count;
+
+    // If the number of features to select is greater than half the total number of features
+    // we will instead randomly select features to *exclude* from the output layer
+    bool invertSelection = number > count / 2;
+    if ( invertSelection )
+      number = count - number;
+
+    // Shuffle <number> features at the start of the iterator
+    auto cursor = allFeats.begin();
+    while ( number-- )
     {
-      break;
+      if ( feedback->isCanceled() )
+        return QVariantMap();
+
+      // Update the distribution to match the number of unshuffled features
+      fidsDistribution.param( std::uniform_int_distribution<int>::param_type( 0, nb - 1 ) );
```

Comment:
on qt 6 builds this would need to be <std::size_t> -- we'll need a conditional #if here to pick the type based on qt version.